### PR TITLE
provider/azure: update AD client

### PR DESCRIPTION
**Pull request by axw**
This updates the Azure Active Directory client, in preparation for a
follow-up which adds the interactive auth-type to the Azure provider. We do
not need to manipulate Applications, only Service Principals; so the
applications client has been removed. The users client has been added so
that we can log the identity of the user that authenticated.

The WithOdataErrorUnlessStatusCode function is a copy of
WithErrorUnlessStatusCode from the autorest/azure package, except that it
expects the error object's name to be "odata.error" instead of just
"error", and does not extract the error message. This is not yet used, and
will be tested in a follow-up.

Requires https://github.com/juju/juju/pull/6270

### CODE REVIEW

File: provider/azure/internal/ad/client.go [hunk 1]
```diff
@@ -5,7 +5,12 @@
 package ad
 
 import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
 	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
 	"github.com/juju/juju/version"
```
Comment (anastasiamac) [inline provider/azure/internal/ad/client.go]: import blocks... I think Azure would be in the 2nd block :D

File: provider/azure/internal/ad/client.go [hunk 2]
```diff
@@ -31,3 +36,54 @@ func NewManagementClient(baseURI string) ManagementClient {
...
+//
+// If this Responder returns an error, the response body will be replaced with
+// an in-memory reader, which needs no further closing.
+func WithOdataErrorUnlessStatusCode(codes ...int) autorest.RespondDecorator {
+	return func(r autorest.Responder) autorest.Responder {
+		return autorest.ResponderFunc(func(resp *http.Response) error {
+			err := r.Respond(resp)
+			if err == nil && !autorest.ResponseHasStatusCode(resp, codes...) {
+				var oe odataRequestError
+				defer resp.Body.Close()
```
Comment (anastasiamac) [inline provider/azure/internal/ad/client.go]: Please comment why defer to close the Body is behind a conditional.

File: provider/azure/internal/ad/serviceprincipals.go [hunk 3]
```diff
@@ -110,9 +111,115 @@ func (client ServicePrincipalsClient) ListResponder(resp *http.Response) (result
...
 		client.ByInspecting(),
-		azure.WithErrorUnlessStatusCode(http.StatusOK),
+		WithOdataErrorUnlessStatusCode(http.StatusOK),
+		autorest.ByUnmarshallingJSON(&result),
+		autorest.ByClosing())
+	result.Response = autorest.Response{Response: resp}
+	return
+}
+
+func (client ServicePrincipalsClient) ListPasswordCredentials(objectId string) (result PasswordCredentialsListResult, err error) {
```
Comment (anastasiamac) [inline provider/azure/internal/ad/serviceprincipals.go]: Comments here and on other calls would be helpful \o/

File: provider/azure/internal/ad/users.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
+// NOTE(axw) this file contains a client for a subset of the
```
Comment (anastasiamac) [inline provider/azure/internal/ad/users.go]: Please add copyright statement :D

File: provider/azure/internal/ad/serviceprincipals.go (hunk 3 above)
Comment (axw) [inline provider/azure/internal/ad/serviceprincipals.go]: This code will be replaced as soon as the Azure SDK is updated. The SDK is auto-generated, and its comments are pretty awful, so I'm not going to spend time on something that'll become irrelevant anyway.

File: provider/azure/internal/ad/client.go (hunk 1 above)
Comment (axw) [inline provider/azure/internal/ad/client.go]: Done.

File: provider/azure/internal/ad/client.go (hunk 2 above)
Comment (axw) [inline provider/azure/internal/ad/client.go]: The whole thing is a straight-up clone of what's in go-autorest, with a very small change that is now documented.

Which reminds me, I need to add the go-autorest LICENSE file.

File: provider/azure/internal/ad/users.go (hunk 4 above)
Comment (axw) [inline provider/azure/internal/ad/users.go]: I've added the LICENSE file from go-autorest, and also added a comment at the top of each file stating that the code is based on that.

**Comment by axw** [comment]
$$merge$$

**Pull request merged**